README.md
# Cake.MsDeploy

Cake.MsDeploy is an Addin that extends [Cake](http://cakebuild.net/) for executing commands with the MsDeploy.exe (Web Deploy) command line interface (cli). 
In order to use this extension, [MsDeploy v3.6](https://www.microsoft.com/en-us/download/details.aspx?id=43717) will already have to be installed on the computer the cake build script is being executed on.  

Release notes can be found [here](ReleaseNotes.md).

## Build Status
Continuous Integration is provided by [AppVeyor](https://www.appveyor.com).  
The build can be found at [https://ci.appveyor.com/project/cakecontrib/cake-msdeploy](https://ci.appveyor.com/project/cakecontrib/cake-msdeploy).

![AppVeyor](https://ci.appveyor.com/api/projects/status/github/cake-contrib/Cake.MsDeploy)

## Referencing

You can reference Cake.MsDeploy in your build script as a cake addin:
```cake
#addin "Cake.MsDeploy"
```  

or nuget reference:

```cake
#addin "nuget:https://www.nuget.org/api/v2?package=Cake.MsDeploy"
```

## Usage

```csharp
#addin "Cake.MsDeploy"

var target = Argument("target", "Default");

Task("MsDeploy-Package")
  .Does(() =>
{
    MsDeploy(new MsDeploySettings
    {
        Verb = Operation.Sync,
        RetryAttempts = 5,
        RetryInterval = 5000,
        Source = new PackageProvider
        {
            Direction = Direction.source,
            Path = "./src/Application.zip"
        },
        Destination = new AutoProvider
        {
            Direction = Direction.dest,
            IncludeAcls = false,
            AuthenticationType = AuthenticationScheme.NTLM,
            ComputerName = "cake.computerName.com",
            TempAgent = true
        },
        AllowUntrusted = true,
        PreSyncCommand = "%windir%\\System32\\inetsrv\\appcmd.exe stop APPPOOL NameOfAppPool",
        PostSyncCommand = "%windir%\\System32\\inetsrv\\appcmd.exe start APPPOOL NameOfAppPool",
        WhatIf = true,
        SetParams = new List<SetParameter>
        {
            new SetParameter
            {
                Name = "IIS Web Application Name",
                Value = "www.cake.com"
            }
        }
    });
});

RunTarget(target);
```

## Documention

Please visit the Cake Documentation site for a list of available aliases:  
[http://cakebuild.net/dsl/msdeploy](http://cakebuild.net/dsl/msdeploy)

## Tests

Cake.MsDeploy is covered by set of xUnit tests.

## Contribution GuideLines

[https://github.com/cake-build/cake/blob/develop/CONTRIBUTING.md](https://github.com/cake-build/cake/blob/develop/CONTRIBUTING.md)

## License

Copyright (c) 2016 Cake Contributions Organization  

Cake.MsDeploy is provided as-is under the MIT license. For more information see [LICENSE](https://github.com/cake-contrib/Cake.MsDeploy/blob/master/LICENSE).

ReleaseNotes.md
## New in 0.1.8 (Released 12/07/2016)
- Update nuspec release notes
- Minor documentation fixes

## New in 0.1.7 (Released 12/07/2016)
- Update Cake.Core references to 0.17.0
- Update Cake build script to run tests before package task

## New in 0.1.6 (Released 12/06/2016)
- Fixed issue with presync and postsync command
- Fixed reference issue with msdeploy example alias

## New in 0.1.5 (Released 11/24/2016)
- Fixed #8 SkipDirectives are never appended to the arguments - thanks @calaa
- Added link to release notes on main ReadMe file
- Added link to CI build project on AppVeyor
- Update license with cake-contrib name

## New in 0.1.4 (Released 11/24/2016)
- Fixed Unrecognized argument 'verb:sync. All arguments must begin with " error - thanks @ecampidoglio
- Fixed link to reference to Cake Documentation website link
- Add Contributions guide link
- Tweaked build.cake script to deploy to nuget on tag commits only.

## New in 0.1.2 (Released 11/06/2016)
- Transfer repository to Cake Contributions Organization
- Update license file to cake-contrib
- Update nuspec file with cake-contrib

## New in 0.1.1 (Released 11/05/2016)
- Fixed documentation with MsDeployAliases example
- Only deploys to nuget.org when pushed on master branch

## New in 0.1.0 (Released 11/04/2016)
- First release of Cake.MsDeploy by Louis Fischer

src/Cake.MsDeploy/MsDeployAliases.cs
using Cake.Core;
using Cake.Core.Annotations;
using System;

namespace Cake.MsDeploy
{
    /// <summary>
    /// <para>
    /// Contains functionality for working with the MsDeploy.exe (Web Deploy) command line interface (cli).</para>
    /// <para>
    /// MsDeploy v3.6 will already have to be installed on the machine the Cake Script is being executed on.
    /// </para>    
    /// See also https://www.microsoft.com/en-us/download/details.aspx?id=43717 
    /// </summary>
    [CakeAliasCategory("MsDeploy")]
    [CakeAliasCategory("Deployment")]
    [CakeNamespaceImport("Cake.MsDeploy")]
    [CakeNamespaceImport("Cake.MsDeploy.Directives")]
    [CakeNamespaceImport("Cake.MsDeploy.Providers.MsDeployProviders")]
    [CakeNamespaceImport("Cake.MsDeploy.Rules")]
    [CakeNamespaceImport("Cake.MsDeploy.Parameters")]
    [CakeNamespaceImport("Cake.MsDeploy.Providers")]
    public static class MsDeployAliases
    {
        /// <summary>
        /// Executes an MsDeploy operation for the given source, destination (optional), and operation settings.
        /// </summary>
        /// <param name="context">The context.</param>
        /// <param name="settings">The MSDeploy settings</param>
        /// <example>
        /// <code>
        ///     var settings = new MsDeploySettings
        ///     {
        ///         Verb = Operation.Sync,
        ///         RetryAttempts = 5,
        ///         RetryInterval = 5000,
        ///         Source = new PackageProvider
        ///         {
        ///             Direction = Direction.source,
        ///             Path = "./src/Application.zip"
        ///         },
        ///         Destination = new AutoProvider
        ///         {
        ///             Direction = Direction.dest,
        ///             IncludeAcls = false,
        ///             AuthenticationType = AuthenticationScheme.NTLM,
        ///             ComputerName = "cake.computerName.com",
        ///             TempAgent = true
        ///          },
        ///          AllowUntrusted = true,
        ///          PreSyncCommand = "%windir%\\System32\\inetsrv\\appcmd.exe stop APPPOOL NameOfAppPool",
        ///          PostSyncCommand = "%windir%\\System32\\inetsrv\\appcmd.exe start APPPOOL NameOfAppPool",
        ///          WhatIf = true,
        ///          SetParams = new List&lt;SetParameter&gt;
        ///          {
        ///             new SetParameter
        ///             {
        ///                 Name = "IIS Web Application Name",
        ///                 Value = "www.cake.com"
        ///             }
        ///          }
        ///      };
        ///
        ///     MsDeploy(settings);
        /// </code>
        /// </example>
        [CakeMethodAlias]
        public static void MsDeploy(this ICakeContext context, MsDeploySettings settings)
        {
            if (context == null)
                throw new ArgumentNullException(nameof(context));

            if (settings == null)
                throw new ArgumentNullException(nameof(settings));

            var runner = new MsDeployRunner(context.FileSystem, context.Environment, context.ProcessRunner, context.Tools);
            runner.Run(settings);
        }
    }
}
